nx/lib/nx/defn/expr.ex

```diff
@@ -129,14 +132,29 @@ defmodule Nx.Defn.Expr do
   @doc """
   Creates a `cond` tensor expression.
   """
+
   def cond([], last) do
-    last
+    Nx.devectorize(last, keep_names: false)
   end
 
-  def cond(clauses, last = out) do
+  def cond(clauses, last) do
     {preds, exprs} = Enum.unzip(clauses)
     {preds, context} = to_exprs(preds)
 
+    [out | _] = broadcasted = Nx.broadcast_vectors([last | exprs])
+
+    [last | exprs] =
+      Enum.zip_with(broadcasted, [last | exprs], fn
+        _expr, original
+        when is_number(original)
+        when original in [:infinity, :neg_infinity, :nan]
+        when is_struct(original, Complex) ->
+          original
+
+        expr, _original ->
+          Nx.devectorize(expr, keep_names: false)
+      end)
+
```

Comment:
This doesn't support keeping scalars inside containers yet. I'm not sure how to "zip" containers or to fit a flat list back into a container.

Reply:
I think this is in the wrong place. The code right below is doing flattening and broadcasting and raising a good error message in case of mismatched clauses. Perhaps we could move the devectorization logic in there instead?